File: CSharp/Client/FromProtobuf.cs
using System;
using System.Collections.Generic;
using System.IO;
using ProtoBuf;
using ThingModel.Proto;

namespace ThingModel.Client
{
    public class FromProtobuf
    {
        private readonly Dictionary<Proto.PropertyType.Type, Type> _prototypesBinding = new Dictionary
            <Proto.PropertyType.Type, Type>
            {
                {Proto.PropertyType.Type.LOCATION, typeof (Property.Location)},
                {Proto.PropertyType.Type.STRING, typeof (Property.String)},
                {Proto.PropertyType.Type.DOUBLE, typeof (Property.Double)},
                {Proto.PropertyType.Type.INT, typeof (Property.Int)},
                {Proto.PropertyType.Type.BOOLEAN, typeof (Property.Boolean)},
                {Proto.PropertyType.Type.DATETIME, typeof (Property.DateTime)},
            };

        protected readonly IDictionary<int,string> StringDeclarations = new Dictionary<int, string>();

        protected string KeyToString(int key)
        {
            if (key == 0)
            {
                return "";
            }

            string value;
            return StringDeclarations.TryGetValue(key, out value) ? value : "undefined";
        }

        protected Wharehouse Wharehouse;

        public FromProtobuf(Wharehouse wharehouse)
        {
            Wharehouse = wharehouse;
        }

        private readonly MemoryStream _memoryInput = new MemoryStream();

        public string Convert(byte[] data, bool check = false)
        {
            _memoryInput.Write(data, 0, data.Length);
            _memoryInput.Position = 0;
            var transaction = Serializer.Deserialize<Transaction>(_memoryInput);
            
            _memoryInput.SetLength(0);

            return Convert(transaction, check);
        }

        public string Convert(Transaction transaction, bool check = false)
        {
            transaction.string_declarations.ForEach(ConvertStringDeclaration);
            transaction.things_remove_list.ForEach(ConvertDelete);
            transaction.thingtypes_declaration_list.ForEach(ConvertThingTypeDeclaration);

            var thingsToConnect = new List<Tuple<Thing, Proto.Thing>>();
            
            foreach (var thing in transaction.things_publish_list)
            {
                var modelThing = ConvertThingPublication(thing, check);
                
                if (thing.connections_change)
                {
                    thingsToConnect.Add(new Tuple<Thing, Proto.Thing>(modelThing, thing));
                }
            }

            foreach (var tuple in thingsToConnect)
            {
                tuple.Item1.Detach();

                foreach (var connection in tuple.Item2.connections)
                {
                    var t = Wharehouse.GetThing(KeyToString(connection));
                    if (t != null)
                    {
                        tuple.Item1.Connect(t);    
                    }
                    
                }
            }

            var senderId = KeyToString(transaction.string_sender_id);

            return senderId;
        }

        protected void ConvertStringDeclaration(StringDeclaration declaration)
        {
            StringDeclarations[declaration.key] = declaration.value;
        }

        public void ConvertDelete(int thingKey)
        {
            var id = KeyToString(thingKey);
            Wharehouse.RemoveThing(Wharehouse.GetThing(id));
        }

        public void ConvertThingTypeDeclaration(Proto.ThingType thingType)
        {
            var modelType = new ThingType(KeyToString(thingType.string_name));
            modelType.Description = KeyToString(thingType.string_description);

            foreach (var propertyType in thingType.properties)
            {
                var type = _prototypesBinding[propertyType.type];
                var modelProperty = new PropertyType(KeyToString(propertyType.string_key),
                                                        type);
                modelProperty.Description = KeyToString(propertyType.string_description);
                modelProperty.Name = KeyToString(propertyType.string_name);
                modelProperty.Required = propertyType.required;

                modelType.DefineProperty(modelProperty);
            }
            Wharehouse.RegisterType(modelType);
        }

        public Thing ConvertThingPublication(Proto.Thing thing, bool check)
        {
            ThingType type = null;
            if (thing.string_type_name != 0)
            {
                // Here the type can be null if the string type name is not registered
                // in the wharehouse (can be an error from a client)

                // So, if the type is null, it rolls back to the default type
                // It's a robust model
                type = Wharehouse.GetThingType(KeyToString(thing.string_type_name));
            }

            var id = KeyToString(thing.string_id);
            
            var modelThing = Wharehouse.GetThing(id);
               
            if (modelThing == null || modelThing.Type != type)
            {
                modelThing = new Thing(KeyToString(thing.string_id), type);
            }

            foreach (var property in thing.properties)
            {
                Property modelProperty = null;
                string key = KeyToString(property.string_key);
                Location location = null;

                switch (property.type)
                {
                    case Proto.Property.Type.LOCATION_POINT:
                        location = new Location.Point();

                        // C# imposes a break statements when a case is not empty
                        // So in 2014 I am using a goto instruction
                        // for the first time in my life
                        goto case Proto.Property.Type.LOCATION_EQUATORIAL;
                        
                    case Proto.Property.Type.LOCATION_LATLNG:
                        location = new Location.LatLng();
                        goto case Proto.Property.Type.LOCATION_EQUATORIAL;

                    case Proto.Property.Type.LOCATION_EQUATORIAL:
                        if (location == null)
                        {
                            location = new Location.Equatorial();
                        }
                        
                        if (property.location_value != null)
                        {

                            location.X = property.location_value.x;
                            location.Y = property.location_value.y;
                            
                            if (!property.location_value.z_null)
                            {
                                location.Z = property.location_value.z;
                            }
                            
                        }

                        modelProperty = new Property.Location(key, location);
                        break;

                    case Proto.Property.Type.STRING:
                        string value = property.string_value.value;
                        if (String.IsNullOrEmpty(value) && property.string_value.string_value != 0)
                        {
                            value = KeyToString(property.string_value.string_value);
                        }
                        modelProperty = new Property.String(key, value);
                        break;

                    case Proto.Property.Type.DOUBLE:
                        modelProperty = new Property.Double(key, property.double_value);
                        break;

                    case Proto.Property.Type.INT:
                        modelProperty = new Property.Int(key, property.int_value);
                        break;

                    case Proto.Property.Type.BOOLEAN:
                        modelProperty = new Property.Boolean(key, property.boolean_value);
                        break;

                    case Proto.Property.Type.DATETIME:
                        modelProperty = new Property.DateTime(key, new DateTime(property.datetime_value));
                        break;
                }
                
                modelThing.SetProperty(modelProperty);
                
            }

            if (check && type != null && !type.Check(modelThing))
            {
                Console.WriteLine("Object «" + id + "» not valid, ignored");
            }
            else
            {
                Wharehouse.RegisterThing(modelThing, false);    
            }
            
            return modelThing;
        }
    }
}

File: Java/src/main/java/org/thingmodel/proto/FromProtobuf.java
package org.thingmodel.proto;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

import org.thingmodel.Location;
import org.thingmodel.Property;
import org.thingmodel.Thing;
import org.thingmodel.Warehouse;
import org.thingmodel.proto.ProtoPropertyType.PropertyType;
import org.thingmodel.proto.ProtoStringDeclaration.StringDeclaration;
import org.thingmodel.proto.ProtoThingType.ThingType;
import org.thingmodel.proto.ProtoTransaction.Transaction;

public class FromProtobuf {
	private static HashMap<PropertyType.Type, Class<? extends Property>> _prototypesBinding = new HashMap<>();
	 
	static {
		_prototypesBinding.put(PropertyType.Type.LOCATION, Property.Location.class);
		_prototypesBinding.put(PropertyType.Type.STRING, Property.String.class);
		_prototypesBinding.put(PropertyType.Type.DOUBLE, Property.Double.class);
		_prototypesBinding.put(PropertyType.Type.INT, Property.Int.class);
		_prototypesBinding.put(PropertyType.Type.BOOLEAN, Property.Boolean.class);
		_prototypesBinding.put(PropertyType.Type.DATETIME, Property.DateTime.class);
	}
	 
	
	private Map<Integer, String> _stringDeclarations = new HashMap<>();
	
	protected String keyToString(int key) {
		if (key == 0) {
			return "";
		}
	
		String value = _stringDeclarations.get(key);
				
		return value != null ? value : "undefined";
	}

	private Warehouse _warehouse;
	
	public FromProtobuf(Warehouse warehouse) {
		_warehouse = warehouse;
	}
	
	public String Convert(Transaction transaction) {
		return Convert(transaction, false);
	}
	
	public String Convert(Transaction transaction, boolean check)
	{
		for (int i = 0, l = transaction.getStringDeclarationsCount(); i < l; ++i) {
			ConvertStringDeclaration(transaction.getStringDeclarations(i));
		}
		
		String senderId = keyToString(transaction.getStringSenderId());
		
		HashSet<Thing> thingsToDelete = new HashSet<>();
		
		for (int i = 0, l = transaction.getThingsRemoveListCount(); i < l; ++i) {
			int key = transaction.getThingsRemoveList(i);
			thingsToDelete.add(_warehouse.getThing(keyToString(key)));
		}
		_warehouse.RemoveCollection(thingsToDelete, senderId);
		
		
		for (int i = 0, l = transaction.getThingtypesDeclarationListCount(); i < l; ++i) {
			ConvertThingTypeDeclaration(transaction.getThingtypesDeclarationList(i), senderId);
		}
		
		class Tuple {
			Thing model;
			org.thingmodel.proto.ProtoThing.Thing proto;
		}
	
		ArrayList<Tuple> thingsToConnect = new ArrayList<>();
		
		for (int i = 0, l = transaction.getThingsPublishListCount(); i < l; ++i) {
			org.thingmodel.proto.ProtoThing.Thing protoThing = transaction.getThingsPublishList(i);
			Thing modelThing = ConvertThingPublication(protoThing, check, senderId);
			
			if (protoThing.getConnectionsChange()) {
				Tuple t = new Tuple();
				t.model = modelThing;
				t.proto = protoThing;
				thingsToConnect.add(t);
			}
		}
		
		for (Tuple tuple : thingsToConnect) {
			tuple.model.DisconnectAll();
			
			for (int i = 0, l = tuple.proto.getConnectionsCount(); i < l; ++i) {
				Thing t = _warehouse.getThing(keyToString(tuple.proto.getConnections(i)));
				
				if (t != null) {
					tuple.model.Connect(t);
				}
			}
			
			_warehouse.RegisterThing(tuple.model, false, false, senderId);
		}
		
		return senderId;
	}

	protected void ConvertStringDeclaration(StringDeclaration declaration) {
		_stringDeclarations.put(declaration.getKey(), declaration.getValue());
	}

	protected void ConvertThingTypeDeclaration(ThingType thingType, String senderId) {
		org.thingmodel.ThingType modelType = new org.thingmodel.ThingType(
				keyToString(thingType.getStringName()));
		modelType.Description = keyToString(thingType.getStringDescription());
		
		for (PropertyType propertyType : thingType.getPropertiesList()) {
			Class<? extends Property> type = _prototypesBinding.get(propertyType.getType());
			String key = keyToString(propertyType.getStringKey());
			
			
			org.thingmodel.PropertyType modelProperty = new org.thingmodel.PropertyType(key, type);
			
			modelProperty.Name = keyToString(propertyType.getStringName());
			modelProperty.Description = keyToString(propertyType.getStringDescription());
			modelProperty.Required = propertyType.getRequired();
		
			modelType.DefineProperty(modelProperty);
		}
		
		_warehouse.RegisterType(modelType, true, senderId);
		
	}

	private Thing ConvertThingPublication(
			org.thingmodel.proto.ProtoThing.Thing thing, boolean check, String senderId) {
		org.thingmodel.ThingType type = null;
	
		if (thing.getStringTypeName() != 0) {
			type = _warehouse.getThingType(keyToString(thing.getStringTypeName()));
		}
		
		String id = keyToString(thing.getStringId());
		
		Thing modelThing = _warehouse.getThing(id);
		
		if (modelThing == null || (
			modelThing.getType() == null && type != null ||
			type == null && modelThing.getType() != null ||
			(modelThing.getType() != null && type != null &&
				modelThing.getType().getName().equals(type.getName())))) {
			modelThing = new Thing(id, type);
		}
		
		for (int i = 0, l = thing.getPropertiesCount(); i < l; ++i) {
			org.thingmodel.proto.ProtoProperty.Property property = thing.getProperties(i);
			
			String key = keyToString(property.getStringKey());
			Property modelProperty = null;
			
			Location location = null;
			
			switch (property.getType()) {
			case LOCATION_POINT:
				location = new Location.Point();
				modelProperty = new Property.Location.Point(key, (Location.Point) location);
			case LOCATION_LATLNG:
				if (location == null) {
					location = new Location.LatLng();
					modelProperty = new Property.Location.LatLng(key, (Location.LatLng) location);
				}
			case LOCATION_EQUATORIAL:
				if (location == null) {
					location = new Location.Equatorial();
					modelProperty = new Property.Location.Equatorial(key, (Location.Equatorial) location);
				}
			
				org.thingmodel.proto.ProtoProperty.Property.Location loc =
						property.getLocationValue();
				
				if (loc != null) {
					location.X = loc.getX();
					location.Y = loc.getY();
					
					if (!loc.getZNull()) {
						location.Z = loc.getZ();
					}
					if (loc.getStringSystem() != 0) {
						location.System = keyToString(loc.getStringSystem());
					}
				}
				
				break;
			case STRING:
				org.thingmodel.proto.ProtoProperty.Property.String sv = property.getStringValue();

                String value;

                if (sv != null) {
                    value = sv.getValue();

                    if (value == null || value.length() == 0 && sv.getStringValue() != 0) {
                        value = keyToString(sv.getStringValue());
                    }
                } else {
                    value = "undefined";
                }

				
				modelProperty = new Property.String(key, value);
				break;
			case BOOLEAN:
				modelProperty = new Property.Boolean(key, property.getBooleanValue());
				break;
			case DATETIME:
				modelProperty = new Property.DateTime(key, new Date(property.getDatetimeValue()));
				break;
			case DOUBLE:
				modelProperty = new Property.Double(key, property.getDoubleValue());
				break;
			case INT:
				modelProperty = new Property.Int(key, property.getIntValue());
				break;
			}
			
			
			modelThing.setProperty(modelProperty);
		}
	
		if (check && type != null && !type.Check(modelThing)) {
			System.out.println("Object «"+id+"» from «"+senderId+"» is not valid, ignored");
		} else if (!thing.getConnectionsChange()) {
			_warehouse.RegisterThing(modelThing, false, false, senderId);
		}
		
		return modelThing;
		
	}

}
